internal/graph/cached_resolver.go

```diff
@@ -0,0 +1,128 @@
+	maxCacheSize int64
+	cacheTTL     time.Duration
+}
+
+var _ CheckResolver = (*CachedCheckResolver)(nil)
+
+// CachedCheckResolverOpt defines an option that can be used to change the behavior of cachedCeckResolver
+// instance.
+type CachedCheckResolverOpt func(*CachedCheckResolver)
+
+// WithMaxCacheSize sets the maximum size of the Check resolution cache. After this
+// maximum size is met, then cache keys will start being evicted with an LRU policy.
+func WithMaxCacheSize(size int64) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.maxCacheSize = size
+	}
+}
+
+// WithCacheTTL sets the TTL (as a duration) for any single Check cache key value.
+func WithCacheTTL(ttl time.Duration) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cacheTTL = ttl
+	}
+}
+
+func WithExistingCache(cache *ccache.Cache[*ResolveCheckResponse]) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cache = cache
+	}
+}
+
+// NewCachedCheckResolver constructs a CheckResolver that delegates Check resolution to the provided delegate,
+// but before delegating the query to the delegate a cache-key lookup is made to see if the Check subproblem
+// has already recently been computed. If the Check subproblem is in the cache, then the response is returned
+// immediately and no recomputation is necessary.
+func NewCachedCheckResolver(delegate CheckResolver, opts ...CachedCheckResolverOpt) *CachedCheckResolver {
+	checker := &CachedCheckResolver{
+		delegate:     delegate,
+		maxCacheSize: defaultMaxCacheSize,
+		cacheTTL:     defaultCacheTTL,
+	}
+	checker.cache = ccache.New(
+		ccache.Configure[*ResolveCheckResponse]().MaxSize(checker.maxCacheSize),
+	)
+
+	for _, opt := range opts {
+		opt(checker)
+	}
+
+	return checker
+}
+
+func (c *CachedCheckResolver) ResolveCheck(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+
+	dispatchedCheckCounter.Inc()
+
+	cacheKey, err := checkRequestCacheKey(req)
+	if err != nil {
+		return nil, fmt.Errorf("cache key computation failed with error: %w", err)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	if cachedResp != nil && !cachedResp.Expired() {
+		dispatchedCheckCachedCounter.Inc()
+		return cachedResp.Value(), nil
+	}
+
+	resp, err := c.delegate.ResolveCheck(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	c.cache.Set(cacheKey, resp, c.cacheTTL)
+	return resp, nil
+}
+
+// checkRequestCacheKey converts the ResolveCheckRequest into a canonical cache key that can be
+// used for Check resolution cache key lookups.
+//
+// We use the 'gob' package to produce a determinstic byte ordering for the contextual tuples provided
+// in the request body. The same tuple provided with the same contextual tuples should produce the same
+// cache key.
+func checkRequestCacheKey(req *ResolveCheckRequest) (string, error) {
+
+	var b bytes.Buffer
+	if err := gob.NewEncoder(&b).Encode(req.GetTupleKey()); err != nil {
```

Comment:
Verify that this will produce a stable encoding even if the underlying protobuf message evolves.

We may need to generate a canonical representation of a TupleKey explicitly by serializing the `(object, relation, user)` bit and concatenating that with some deterministic and canonical representation of the slice of ContextualTuples (maybe we sort the slice and then .


See:
https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510
https://github.com/golang/protobuf/issues/1121


Also,

> Proto serialization is not stable, thus you can't rely on marshalling and hashing the output to produce the same hash for the same message.

From https://developers.google.com/protocol-buffers/docs/reference/go/faq#hash

> How do I use a protocol buffer message as a hash key?

> You need canonical serialization, where the marshaled output of a protocol buffer message is guaranteed to be stable over time. Unfortunately, no specification for canonical serialization exists at this time. You'll need to write your own or find a way to avoid needing one.

Reply:
Interesting..

Reply:
See https://protobuf.dev/getting-started/cpptutorial/#parsing-serialization

<img width="1051" alt="image" src="https://github.com/openfga/openfga/assets/5374887/daa43ac1-9314-4f01-8816-953283fbd4ea">

We should (in another PR that is a pre-requisite of this one) wrap the `openfgav1.CheckRequest` object into a `openfgav1.ExtendedCheckRequest` that defines the `Equals` method.

I've been wanting to do something similar with `openfgav1.AuthorizationModel`. Since `Typesystem` is basically an extended AuthorizationModel with validation and parsing methods, I've been wanting to stop referencing `openfgav1.AuthorizationModel` in our codebase and replace it with `Typesystem`

Reply:
We should do that - however, I think it is orthogonal to the caching work here.  Let's not bloat the scope of this PR.

Reply:
Tracked via https://github.com/openfga/openfga/issues/925